feat(food): new module

xDivisionByZerox opened the pull request:
## Description

MVP implementation for a new `FoodModule`. The locale data is based on the dataset of [the ruby Faker](https://github.com/faker-ruby/faker/blob/main/lib/locales/en/food.yml) library.

Feel free to discuss certain implementation details or locale data as you wish.

## More Context

Resolves #1593.

xDivisionByZerox committed:
feat(food): new module

xDivisionByZerox committed:
feat(locale): add food data for en

xDivisionByZerox committed:
test(food): add

xDivisionByZerox committed:
docs: add food to api page

xDivisionByZerox committed:
docs: add basic JSDocs

ST-DDT commented on test.ts:
This file seems to have been committed by accident.

xDivisionByZerox replied:
Wups. was testing something. Thank you.

ST-DDT commented on test/modules/__snapshots__/food.spec.ts.snap:
```diff
@@ -0,0 +1,43 @@
+
+exports[`food > 42 > vegetable 1`] = `"Dried Chinese Broccoli"`;
+
+exports[`food > 1211 > description 1`] = `"A special turquoise Sunflower Seeds from San Marino. To support the strong flavor it is sided with a tablespoon of Chillies."`;
```
🤣

xDivisionByZerox replied:
Glad you had a good laught. Was my intention with these description. Hope it's still "professional" enougth.

ST-DDT commented on test/modules/food.spec.ts:
```diff
@@ -0,0 +1,71 @@
+
+describe('food', () => {
+  seededTests(faker, 'food', (t) => {
+    t.it('description');
+
+    t.it('dish');
+
+    t.it('ethnicCategory');
+
+    t.it('fruit');
+
+    t.it('ingredient');
+
+    t.it('spice');
+
+    t.it('vegetable');
```
You could merge them together, by just calling `t.itEach(...)`

xDivisionByZerox replied:
I know that this is a possibility, but had trouble with the auto completion on the keys 🤔

xDivisionByZerox committed:
chore: remove accidental committed file

matthewmayer commented on src/locales/en/food/dish.ts:
```diff
@@ -0,0 +1,56 @@
+export default [
+  'Arepas',
```
I feel the dish definitions are very limited, given this is probably the most useful method for a lot of use cases (eg restaurant menus). I think we could add a lot of extra dishes from the Ruby version and/or use fake patterns to allow for many more dish names.

xDivisionByZerox replied:
Agreed. 

---

An additional thought: Should dishes be categorized? Like having a file for categories:

- pizza: Pizza Margherita, Pizza Pepperoni, New York Style Pizza, etc
- pasta: Lasagna, Spaghetti, Tagliatelle, etc
- salad: chef salad, caesar salad, potato salad

These files do not have to be exposed individually but are exposed via a fake expression in the `dish.ts` file. That why each locale can have different dish categories.

ST-DDT replied:
While I think, that we should extend the list a bit, I don't see a need for fake patterns here.

> An additional thought: Should dishes be categorized? Like having a file for categories:

If there is sufficient user demand for it maybe. Not as part of the MVP.

matthewmayer replied:
I had a few use case of needing to generate menus for hundreds of restaurants each of which would have 20-50 menu items. 

Fake patterns would allow for 1000s of unique dishes without needing 1000s of definitions. Eg if you have "{{food.ethnicCategory}} Salmon" or "{{food.fruit}} Pie" in the list you multiply your unique dishes dramatically.

ST-DDT replied:
**Team Decision**

- We will add both `dish` and `dish_patterns`.

matthewmayer commented on src/locales/en/food/fruit.ts:
```diff
@@ -0,0 +1,71 @@
+export default [
+  'Apples',
```
We should be consistent if these are singular or plural

xDivisionByZerox replied:
Any preferences? I feel like these should be sigular, but can't give reasoning for that.

ST-DDT replied:
Just go for singular for now. Most plurals are easily achievable by just appending an `s` to the pattern.
Should they be lowercase as well?

matthewmayer replied:
Lowercase and singular seems most flexible in English

matthewmayer replied:
Singular also allows for phrases like "{{food.fruit}} sauce"

ejcheng replied:
I would also be in favor of lowercase and singular

matthewmayer commented on src/locales/en/food/description_pattern.ts:
```diff
@@ -0,0 +1,9 @@
+  '{{food.ingredient}} with a pinch of {{food.spice}}, topped by a caramelized {{food.fruit}} with whipped cream',
+  'A slow-roasted {{animal.bird}} with a crisp, golden exterior. Stuffed with {{food.fruit}} and covered in {{food.fruit}} sauce. Sides with {{food.vegetable}} puree and wild {{food.vegetable}}.',
+  'One slice of exotic {{animal.crocodilia}} meat, {{food.spice}} {{food.spice}} butter, with an additional side.',
+  '28-day aged 300g {{animal.cow}} steak, with choice of two sides.',
```
Many languages including English distinguish between animal words and meat words. 

Eg beef, lamb, venison vs cow, sheep, deer

I think we could have a food.protein or food.meat method which returns popular meats or meat substitutes which can be used in other patterns eg chicken,beef,lamb,venison,veal,turkey,tofu

matthewmayer commented on src/locales/en/food/description_pattern.ts:
```diff
@@ -0,0 +1,9 @@
+export default [
+  'Three {{food.ingredient}} with {{food.vegetable}}, {{food.vegetable}}, {{food.vegetable}}, {{food.vegetable}} and {{food.ingredient}}. With a side of baked {{food.fruit}}, and your choice of {{food.ingredient}} or {{food.ingredient}}.',
+  '{{food.ingredient}} with a pinch of {{food.spice}}, topped by a caramelized {{food.fruit}} with whipped cream',
+  'A slow-roasted {{animal.bird}} with a crisp, golden exterior. Stuffed with {{food.fruit}} and covered in {{food.fruit}} sauce. Sides with {{food.vegetable}} puree and wild {{food.vegetable}}.',
```
could also be useful to have a food.adjective which is a list of adjectives commonly applied to food that can be used in dish names and descriptions, e.g.
```
[
    "crunchy",
    "delicious",
    "golden",
    "savory",
    "spicy",
    "sweet",
    "bitter",
    "tangy",
    "creamy",
    "juicy",
    "crispy",
    "tender",
    "salty",
    "sour",
    "zesty",
    "fluffy",
    "moist",
    "fresh",
    "rich",
    "smoky"
  ]
}
```

ST-DDT commented:
Please fix the "merge conflicts" (although I don't think there are any for local executions).

matthewmayer commented:
@xDivisionByZerox would you be OK with me continuing work on this to address some of the feedback and get to an MVP?

xDivisionByZerox commented:
> @xDivisionByZerox would you be OK with me continuing work on this to address some of the feedback and get to an MVP?

@matthewmayer I would be happy if you would continue working on this.

matthewmayer committed:
fix merge

matthewmayer committed:
new definitions and methods

- add new definitions for adjective, dish_pattern and meat
- add more description patterns
- make most definitions lower case and singular
- alphabeticize definitions

matthewmayer committed:
fix lint

matthewmayer committed:
fix tests

matthewmayer commented:
You can see a sample menu from the Faker Bistro here ;)

https://gist.github.com/matthewmayer/794406e8281e2c796e70495a1598ec50

Generated with:
```js
console.log(faker.helpers.multiple(a=>faker.helpers.fake("**{{food.dish}}** - {{commerce.price({\"max\":30})}}\n\n{{food.description}}\n\n"),{count:50}).join(""))
```

matthewmayer committed:
tweak descriptions

matthewmayer committed:
fix snapshot

ST-DDT commented on src/modules/food/index.ts:
```diff
@@ -0,0 +1,144 @@
+ * To generate the name of a dish, use [dish()](https://fakerjs.dev/api/food.html#dish) and to generate a long description for a dish use [`description()`](https://fakerjs.dev/api/food.html#description). Note that these will not correspond with each other.
+ * You can also generate individual components of a dish such as [spices](https://fakerjs.dev/api/food.html#spice), [vegetables](https://fakerjs.dev/api/food.html#vegetable), [meats](https://fakerjs.dev/api/food.html#meat), [fruits](https://fakerjs.dev/api/food.html#fruit), or generic [ingredients](https://fakerjs.dev/api/food.html#ingredient).
+ */
+export class FoodModule {
+  constructor(private readonly faker: Faker) {
+    bindThisToMemberFunctions(this);
+  }
```
```suggestion
export class FoodModule extends ModuleBase {
```

ST-DDT commented on src/modules/food/index.ts:
```diff
@@ -0,0 +1,144 @@
+  dish(): string {
+    // A 50/50 mix of specific dishes and dish_patterns
+    const toTitleCase = (d: string) =>
+      d.toLowerCase().replace(/\b\w/g, (s) => s.toUpperCase());
```
I would probably omit the toLowercase here in case we have parts that are already capitalized e.g.

`MC Danold's delicious beacon` -> `Mc Danold'S Delicious Beacon`

Also the `'S` should probably be not capitalized.

If we capitalize the dishes anyway, should we also capitalize them in the locale data?

matthewmayer replied:
No, because we don't want them capitalised in dish descriptions.

matthewmayer replied:
replaced with a less aggressive capitalization function

matthewmayer committed:
Update index.ts

Co-authored-by: ST-DDT <ST-DDT@gmx.de>

matthewmayer committed:
fix imports

matthewmayer committed:
fix lint, tweak title case

matthewmayer committed:
Merge branch 'next' into feat/food-module

matthewmayer committed:
MD tweak

matthewmayer committed:
Merge branch 'next' into feat/food-module

ST-DDT commented:
FFR: This feature would benefit from https://github.com/faker-js/faker/issues/2584

- https://github.com/faker-js/faker/issues/2584#issuecomment-1936971449

ST-DDT commented on src/modules/food/index.ts:
```diff
@@ -0,0 +1,141 @@
+   * Generates a random dish description.
+   *
+   * @example
+   * faker.food.description() // ''An exquisite ostrich roast, infused with the essence of longan, slow-roasted to bring out its natural flavors and served with a side of creamy red cabbage'
```
```suggestion
   * faker.food.description() // 'An exquisite ostrich roast, infused with the essence of longan, slow-roasted to bring out its natural flavors and served with a side of creamy red cabbage'
```

ST-DDT reviewed (dismissed):
A little typo but otherwise good to go.

matthewmayer committed:
Update index.ts

Co-authored-by: ST-DDT <ST-DDT@gmx.de>

ST-DDT committed:
Merge branch 'next' into feat/food-module

IgnusG commented:
Just checking out of curiosity, I see a new version with this module has not been published yet. Is that intentional?

matthewmayer commented:
Yes it's intentional it will appear in v9 but there are a few more things to merge before the first alpha/beta version of v9 will be released.
